File: kits/data-integration/providers/connector_protocol/rest.ts
// REST — connector_protocol provider
// Generic REST API connector with pagination, auth, rate limiting, and retry logic

export interface ConnectorConfig {
  baseUrl?: string;
  connectionString?: string;
  auth?: Record<string, unknown>;
  headers?: Record<string, string>;
  options?: Record<string, unknown>;
}

export interface QuerySpec {
  path?: string;
  query?: string;
  params?: Record<string, unknown>;
  cursor?: string;
  limit?: number;
}

export interface WriteResult { created: number; updated: number; skipped: number; errors: number; }
export interface TestResult { connected: boolean; message: string; latencyMs?: number; }
export interface StreamDef { name: string; schema: Record<string, unknown>; supportedSyncModes: string[]; }
export interface DiscoveryResult { streams: StreamDef[]; }

export const PROVIDER_ID = 'rest';
export const PLUGIN_TYPE = 'connector_protocol';

type PaginationStyle = 'cursor' | 'offset' | 'page' | 'link';
type AuthStyle = 'bearer' | 'api_key' | 'oauth2' | 'basic';

interface AuthConfig {
  style: AuthStyle;
  token?: string;
  apiKey?: string;
  apiKeyHeader?: string;
  clientId?: string;
  clientSecret?: string;
  tokenUrl?: string;
  username?: string;
  password?: string;
}

function buildAuthHeaders(auth: AuthConfig): Record<string, string> {
  switch (auth.style) {
    case 'bearer':
      return { Authorization: `Bearer ${auth.token}` };
    case 'api_key':
      return { [auth.apiKeyHeader ?? 'X-API-Key']: auth.apiKey ?? '' };
    case 'basic': {
      const encoded = Buffer.from(`${auth.username}:${auth.password}`).toString('base64');
      return { Authorization: `Basic ${encoded}` };
    }
    case 'oauth2':
      return { Authorization: `Bearer ${auth.token}` };
    default:
      return {};
  }
}

async function refreshOAuth2Token(auth: AuthConfig): Promise<string> {
  if (!auth.tokenUrl || !auth.clientId || !auth.clientSecret) {
    throw new Error('OAuth2 requires tokenUrl, clientId, and clientSecret');
  }
  const resp = await fetch(auth.tokenUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'client_credentials',
      client_id: auth.clientId,
      client_secret: auth.clientSecret,
    }),
  });
  if (!resp.ok) throw new Error(`OAuth2 token refresh failed: ${resp.status}`);
  const data = await resp.json();
  return data.access_token;
}

async function fetchWithRetry(
  url: string,
  init: RequestInit,
  maxRetries = 3
): Promise<Response> {
  let lastError: Error | null = null;
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      const resp = await fetch(url, init);
      if (resp.status === 429 || resp.status >= 500) {
        const retryAfter = resp.headers.get('Retry-After');
        const delayMs = retryAfter
          ? parseInt(retryAfter, 10) * 1000
          : Math.min(1000 * Math.pow(2, attempt), 30000);
        await new Promise(r => setTimeout(r, delayMs));
        continue;
      }
      return resp;
    } catch (err) {
      lastError = err instanceof Error ? err : new Error(String(err));
      if (attempt < maxRetries) {
        await new Promise(r => setTimeout(r, 1000 * Math.pow(2, attempt)));
      }
    }
  }
  throw lastError ?? new Error('Request failed after retries');
}

export class RestConnectorProvider {
  private cachedToken: string | null = null;

  private async getHeaders(config: ConnectorConfig): Promise<Record<string, string>> {
    const headers: Record<string, string> = { 'Content-Type': 'application/json', ...config.headers };
    if (config.auth) {
      const auth = config.auth as unknown as AuthConfig;
      if (auth.style === 'oauth2' && !this.cachedToken) {
        this.cachedToken = await refreshOAuth2Token(auth);
        auth.token = this.cachedToken;
      }
      Object.assign(headers, buildAuthHeaders(auth));
    }
    return headers;
  }

  async *read(query: QuerySpec, config: ConnectorConfig): AsyncGenerator<Record<string, unknown>> {
    const baseUrl = config.baseUrl?.replace(/\/$/, '') ?? '';
    const path = query.path ?? '/';
    const paginationStyle: PaginationStyle = (config.options?.pagination as PaginationStyle) ?? 'offset';
    const pageSize = query.limit ?? 100;
    const dataKey = (config.options?.dataKey as string) ?? 'data';
    const headers = await this.getHeaders(config);

    let cursor = query.cursor;
    let offset = 0;
    let page = 1;
    let nextLink: string | null = null;
    let hasMore = true;

    while (hasMore) {
      let url: string;
      if (nextLink) {
        url = nextLink;
      } else {
        const params = new URLSearchParams();
        if (query.params) {
          for (const [k, v] of Object.entries(query.params)) params.set(k, String(v));
        }
        params.set('limit', String(pageSize));
        if (paginationStyle === 'cursor' && cursor) params.set('cursor', cursor);
        if (paginationStyle === 'offset') params.set('offset', String(offset));
        if (paginationStyle === 'page') params.set('page', String(page));
        url = `${baseUrl}${path}?${params.toString()}`;
      }

      const resp = await fetchWithRetry(url, { method: 'GET', headers });
      if (!resp.ok) throw new Error(`REST read failed: ${resp.status} ${resp.statusText}`);
      const body = await resp.json();

      const records: Record<string, unknown>[] = Array.isArray(body)
        ? body
        : (body[dataKey] ?? []);

      for (const record of records) {
        yield record;
      }

      if (records.length < pageSize) {
        hasMore = false;
      } else if (paginationStyle === 'cursor') {
        cursor = body.next_cursor ?? body.cursor ?? body.nextCursor;
        hasMore = !!cursor;
      } else if (paginationStyle === 'offset') {
        offset += records.length;
      } else if (paginationStyle === 'page') {
        page++;
      } else if (paginationStyle === 'link') {
        nextLink = body.links?.next ?? null;
        hasMore = !!nextLink;
      }
    }
  }

  async write(records: Record<string, unknown>[], config: ConnectorConfig): Promise<WriteResult> {
    const baseUrl = config.baseUrl?.replace(/\/$/, '') ?? '';
    const writePath = (config.options?.writePath as string) ?? '/';
    const batchSize = (config.options?.batchSize as number) ?? 50;
    const method = (config.options?.writeMethod as string) ?? 'POST';
    const headers = await this.getHeaders(config);
    const result: WriteResult = { created: 0, updated: 0, skipped: 0, errors: 0 };

    for (let i = 0; i < records.length; i += batchSize) {
      const batch = records.slice(i, i + batchSize);
      try {
        const resp = await fetchWithRetry(`${baseUrl}${writePath}`, {
          method,
          headers,
          body: JSON.stringify(batch),
        });
        if (resp.ok) {
          const body = await resp.json();
          result.created += body.created ?? batch.length;
          result.updated += body.updated ?? 0;
          result.skipped += body.skipped ?? 0;
        } else {
          result.errors += batch.length;
        }
      } catch {
        result.errors += batch.length;
      }
    }
    return result;
  }

  async test(config: ConnectorConfig): Promise<TestResult> {
    const baseUrl = config.baseUrl?.replace(/\/$/, '') ?? '';
    const healthPath = (config.options?.healthPath as string) ?? '/';
    const headers = await this.getHeaders(config);
    const start = Date.now();
    try {
      const resp = await fetchWithRetry(`${baseUrl}${healthPath}`, { method: 'GET', headers }, 1);
      return {
        connected: resp.ok,
        message: resp.ok ? 'Connected successfully' : `HTTP ${resp.status}`,
        latencyMs: Date.now() - start,
      };
    } catch (err) {
      return { connected: false, message: String(err), latencyMs: Date.now() - start };
    }
  }

  async discover(config: ConnectorConfig): Promise<DiscoveryResult> {
    const baseUrl = config.baseUrl?.replace(/\/$/, '') ?? '';
    const discoveryPath = (config.options?.discoveryPath as string) ?? '/';
    const headers = await this.getHeaders(config);
    try {
      const resp = await fetchWithRetry(`${baseUrl}${discoveryPath}`, { method: 'GET', headers }, 1);
      if (!resp.ok) return { streams: [] };
      const body = await resp.json();
      const endpoints: string[] = Array.isArray(body) ? body : Object.keys(body);
      return {
        streams: endpoints.map(name => ({
          name,
          schema: {},
          supportedSyncModes: ['full_refresh', 'incremental'],
        })),
      };
    } catch {
      return { streams: [] };
    }
  }
}

export default RestConnectorProvider;

File: kits/data-integration/providers/connector_protocol/rest.rs
// REST — connector_protocol provider
// Generic REST API connector with pagination, auth, rate limiting, and retry logic

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::time::{Duration, Instant};

pub const PROVIDER_ID: &str = "rest";
pub const PLUGIN_TYPE: &str = "connector_protocol";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectorConfig {
    pub base_url: Option<String>,
    pub connection_string: Option<String>,
    pub auth: Option<AuthConfig>,
    pub headers: Option<HashMap<String, String>>,
    pub options: Option<HashMap<String, Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthConfig {
    pub style: String,
    pub token: Option<String>,
    pub api_key: Option<String>,
    pub api_key_header: Option<String>,
    pub client_id: Option<String>,
    pub client_secret: Option<String>,
    pub token_url: Option<String>,
    pub username: Option<String>,
    pub password: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QuerySpec {
    pub path: Option<String>,
    pub query: Option<String>,
    pub params: Option<HashMap<String, Value>>,
    pub cursor: Option<String>,
    pub limit: Option<u64>,
}

pub type Record = HashMap<String, Value>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WriteResult {
    pub created: u64,
    pub updated: u64,
    pub skipped: u64,
    pub errors: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TestResult {
    pub connected: bool,
    pub message: String,
    pub latency_ms: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StreamDef {
    pub name: String,
    pub schema: HashMap<String, Value>,
    pub supported_sync_modes: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscoveryResult {
    pub streams: Vec<StreamDef>,
}

#[derive(Debug)]
pub struct ConnectorError(pub String);

impl std::fmt::Display for ConnectorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for ConnectorError {}

fn build_auth_headers(auth: &AuthConfig) -> HashMap<String, String> {
    let mut headers = HashMap::new();
    match auth.style.as_str() {
        "bearer" => {
            if let Some(token) = &auth.token {
                headers.insert("Authorization".into(), format!("Bearer {}", token));
            }
        }
        "api_key" => {
            let header = auth.api_key_header.as_deref().unwrap_or("X-API-Key");
            if let Some(key) = &auth.api_key {
                headers.insert(header.into(), key.clone());
            }
        }
        "basic" => {
            let user = auth.username.as_deref().unwrap_or("");
            let pass = auth.password.as_deref().unwrap_or("");
            let encoded = base64_encode(&format!("{}:{}", user, pass));
            headers.insert("Authorization".into(), format!("Basic {}", encoded));
        }
        "oauth2" => {
            if let Some(token) = &auth.token {
                headers.insert("Authorization".into(), format!("Bearer {}", token));
            }
        }
        _ => {}
    }
    headers
}

fn base64_encode(input: &str) -> String {
    const CHARS: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    let bytes = input.as_bytes();
    let mut result = String::new();
    for chunk in bytes.chunks(3) {
        let b0 = chunk[0] as u32;
        let b1 = if chunk.len() > 1 { chunk[1] as u32 } else { 0 };
        let b2 = if chunk.len() > 2 { chunk[2] as u32 } else { 0 };
        let triple = (b0 << 16) | (b1 << 8) | b2;
        result.push(CHARS[((triple >> 18) & 0x3F) as usize] as char);
        result.push(CHARS[((triple >> 12) & 0x3F) as usize] as char);
        if chunk.len() > 1 {
            result.push(CHARS[((triple >> 6) & 0x3F) as usize] as char);
        } else {
            result.push('=');
        }
        if chunk.len() > 2 {
            result.push(CHARS[(triple & 0x3F) as usize] as char);
        } else {
            result.push('=');
        }
    }
    result
}

pub struct RestConnectorProvider {
    config: Option<ConnectorConfig>,
    client: reqwest::Client,
}

impl RestConnectorProvider {
    pub fn new() -> Self {
        Self {
            config: None,
            client: reqwest::Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .unwrap_or_default(),
        }
    }

    fn build_headers(&self, config: &ConnectorConfig) -> HashMap<String, String> {
        let mut headers: HashMap<String, String> = HashMap::new();
        headers.insert("Content-Type".into(), "application/json".into());
        if let Some(cfg_headers) = &config.headers {
            headers.extend(cfg_headers.clone());
        }
        if let Some(auth) = &config.auth {
            headers.extend(build_auth_headers(auth));
        }
        headers
    }

    async fn fetch_with_retry(
        &self,
        url: &str,
        headers: &HashMap<String, String>,
        max_retries: u32,
    ) -> Result<reqwest::Response, ConnectorError> {
        let mut last_err = String::new();
        for attempt in 0..=max_retries {
            let mut req = self.client.get(url);
            for (k, v) in headers {
                req = req.header(k.as_str(), v.as_str());
            }
            match req.send().await {
                Ok(resp) => {
                    let status = resp.status().as_u16();
                    if status == 429 || status >= 500 {
                        let delay = resp
                            .headers()
                            .get("Retry-After")
                            .and_then(|v| v.to_str().ok())
                            .and_then(|v| v.parse::<u64>().ok())
                            .unwrap_or(2u64.pow(attempt));
                        tokio::time::sleep(Duration::from_secs(delay)).await;
                        continue;
                    }
                    return Ok(resp);
                }
                Err(e) => {
                    last_err = e.to_string();
                    if attempt < max_retries {
                        tokio::time::sleep(Duration::from_secs(2u64.pow(attempt))).await;
                    }
                }
            }
        }
        Err(ConnectorError(format!("Request failed after retries: {}", last_err)))
    }

    pub async fn read(
        &self,
        query: &QuerySpec,
        config: &ConnectorConfig,
    ) -> Result<Vec<Record>, ConnectorError> {
        let base = config.base_url.as_deref().unwrap_or("").trim_end_matches('/');
        let path = query.path.as_deref().unwrap_or("/");
        let page_size = query.limit.unwrap_or(100);
        let headers = self.build_headers(config);
        let data_key = config
            .options.as_ref()
            .and_then(|o| o.get("dataKey"))
            .and_then(|v| v.as_str())
            .unwrap_or("data");
        let pagination = config
            .options.as_ref()
            .and_then(|o| o.get("pagination"))
            .and_then(|v| v.as_str())
            .unwrap_or("offset");

        let mut all_records = Vec::new();
        let mut offset: u64 = 0;
        let mut cursor = query.cursor.clone();

        loop {
            let mut url = format!("{}{}?limit={}", base, path, page_size);
            if let Some(params) = &query.params {
                for (k, v) in params {
                    url.push_str(&format!("&{}={}", k, v));
                }
            }
            match pagination {
                "cursor" => {
                    if let Some(c) = &cursor {
                        url.push_str(&format!("&cursor={}", c));
                    }
                }
                "offset" => url.push_str(&format!("&offset={}", offset)),
                _ => {}
            }

            let resp = self.fetch_with_retry(&url, &headers, 3).await?;
            let body: Value = resp.json().await.map_err(|e| ConnectorError(e.to_string()))?;

            let records: Vec<Record> = if let Some(arr) = body.as_array() {
                arr.iter()
                    .filter_map(|v| serde_json::from_value(v.clone()).ok())
                    .collect()
            } else if let Some(arr) = body.get(data_key).and_then(|v| v.as_array()) {
                arr.iter()
                    .filter_map(|v| serde_json::from_value(v.clone()).ok())
                    .collect()
            } else {
                Vec::new()
            };

            let count = records.len() as u64;
            all_records.extend(records);

            if count < page_size {
                break;
            }
            match pagination {
                "cursor" => {
                    cursor = body
                        .get("next_cursor")
                        .or_else(|| body.get("cursor"))
                        .and_then(|v| v.as_str())
                        .map(String::from);
                    if cursor.is_none() {
                        break;
                    }
                }
                "offset" => offset += count,
                _ => break,
            }
        }
        Ok(all_records)
    }

    pub async fn write(
        &self,
        records: &[Record],
        config: &ConnectorConfig,
    ) -> Result<WriteResult, ConnectorError> {
        let base = config.base_url.as_deref().unwrap_or("").trim_end_matches('/');
        let write_path = config
            .options.as_ref()
            .and_then(|o| o.get("writePath"))
            .and_then(|v| v.as_str())
            .unwrap_or("/");
        let batch_size = config
            .options.as_ref()
            .and_then(|o| o.get("batchSize"))
            .and_then(|v| v.as_u64())
            .unwrap_or(50) as usize;
        let headers = self.build_headers(config);

        let mut result = WriteResult { created: 0, updated: 0, skipped: 0, errors: 0 };

        for chunk in records.chunks(batch_size) {
            let url = format!("{}{}", base, write_path);
            let body = serde_json::to_string(chunk).map_err(|e| ConnectorError(e.to_string()))?;
            let mut req = self.client.post(&url);
            for (k, v) in &headers {
                req = req.header(k.as_str(), v.as_str());
            }
            match req.body(body).send().await {
                Ok(resp) if resp.status().is_success() => {
                    result.created += chunk.len() as u64;
                }
                _ => {
                    result.errors += chunk.len() as u64;
                }
            }
        }
        Ok(result)
    }

    pub async fn test_connection(
        &self,
        config: &ConnectorConfig,
    ) -> Result<TestResult, ConnectorError> {
        let base = config.base_url.as_deref().unwrap_or("").trim_end_matches('/');
        let health_path = config
            .options.as_ref()
            .and_then(|o| o.get("healthPath"))
            .and_then(|v| v.as_str())
            .unwrap_or("/");
        let headers = self.build_headers(config);
        let start = Instant::now();

        match self.fetch_with_retry(&format!("{}{}", base, health_path), &headers, 1).await {
            Ok(resp) => {
                let ok = resp.status().is_success();
                Ok(TestResult {
                    connected: ok,
                    message: if ok { "Connected successfully".into() } else { format!("HTTP {}", resp.status()) },
                    latency_ms: Some(start.elapsed().as_millis() as u64),
                })
            }
            Err(e) => Ok(TestResult {
                connected: false,
                message: e.0,
                latency_ms: Some(start.elapsed().as_millis() as u64),
            }),
        }
    }

    pub async fn discover(
        &self,
        config: &ConnectorConfig,
    ) -> Result<DiscoveryResult, ConnectorError> {
        let base = config.base_url.as_deref().unwrap_or("").trim_end_matches('/');
        let disc_path = config
            .options.as_ref()
            .and_then(|o| o.get("discoveryPath"))
            .and_then(|v| v.as_str())
            .unwrap_or("/");
        let headers = self.build_headers(config);

        match self.fetch_with_retry(&format!("{}{}", base, disc_path), &headers, 1).await {
            Ok(resp) if resp.status().is_success() => {
                let body: Value = resp.json().await.map_err(|e| ConnectorError(e.to_string()))?;
                let names: Vec<String> = if let Some(arr) = body.as_array() {
                    arr.iter().filter_map(|v| v.as_str().map(String::from)).collect()
                } else if let Some(obj) = body.as_object() {
                    obj.keys().cloned().collect()
                } else {
                    Vec::new()
                };
                Ok(DiscoveryResult {
                    streams: names
                        .into_iter()
                        .map(|name| StreamDef {
                            name,
                            schema: HashMap::new(),
                            supported_sync_modes: vec!["full_refresh".into(), "incremental".into()],
                        })
                        .collect(),
                })
            }
            _ => Ok(DiscoveryResult { streams: Vec::new() }),
        }
    }
}
